File: mock/vrp_mock/telnet.py
"""Telnet protocol constants and handling.

Mirrors constants from src-tauri/src/telnet.rs for protocol compatibility.
"""

from enum import IntEnum
from typing import Tuple, List, Optional
from dataclasses import dataclass


# Telnet protocol constants (RFC 854)
class TelnetCmd(IntEnum):
    """Telnet command bytes."""
    SE = 240    # Sub-negotiation End
    NOP = 241   # No Operation
    DM = 242    # Data Mark
    BRK = 243   # Break
    IP = 244    # Interrupt Process
    AO = 245    # Abort Output
    AYT = 246   # Are You There
    EC = 247    # Erase Character
    EL = 248    # Erase Line
    GA = 249    # Go Ahead
    SB = 250    # Sub-negotiation Begin
    WILL = 251
    WONT = 252
    DO = 253
    DONT = 254
    IAC = 255   # Interpret As Command


# Telnet options
class TelnetOpt(IntEnum):
    """Telnet option codes."""
    ECHO = 1
    SUPPRESS_GO_AHEAD = 3
    STATUS = 5
    TIMING_MARK = 6
    TERMINAL_TYPE = 24
    NAWS = 31           # Negotiate About Window Size
    TERMINAL_SPEED = 32
    LINEMODE = 34
    NEW_ENVIRON = 39


@dataclass
class TelnetCommand:
    """Parsed telnet command."""
    cmd: int
    option: Optional[int] = None
    data: Optional[bytes] = None


class TelnetParser:
    """Parse telnet protocol from byte stream."""

    def __init__(self):
        self._state = 'normal'
        self._subneg_option = 0
        self._subneg_data = bytearray()

    def parse(self, data: bytes) -> Tuple[bytes, List[TelnetCommand]]:
        """Parse input bytes, return (clean_data, commands)."""
        output = bytearray()
        commands = []

        for byte in data:
            if self._state == 'normal':
                if byte == TelnetCmd.IAC:
                    self._state = 'iac'
                else:
                    output.append(byte)
            elif self._state == 'iac':
                if byte == TelnetCmd.IAC:
                    output.append(TelnetCmd.IAC)
                    self._state = 'normal'
                elif byte == TelnetCmd.WILL:
                    self._state = 'will'
                elif byte == TelnetCmd.WONT:
                    self._state = 'wont'
                elif byte == TelnetCmd.DO:
                    self._state = 'do'
                elif byte == TelnetCmd.DONT:
                    self._state = 'dont'
                elif byte == TelnetCmd.SB:
                    self._state = 'sb'
                elif byte == TelnetCmd.SE:
                    self._state = 'normal'
                else:
                    self._state = 'normal'
            elif self._state == 'will':
                commands.append(TelnetCommand(TelnetCmd.WILL, byte))
                self._state = 'normal'
            elif self._state == 'wont':
                commands.append(TelnetCommand(TelnetCmd.WONT, byte))
                self._state = 'normal'
            elif self._state == 'do':
                commands.append(TelnetCommand(TelnetCmd.DO, byte))
                self._state = 'normal'
            elif self._state == 'dont':
                commands.append(TelnetCommand(TelnetCmd.DONT, byte))
                self._state = 'normal'
            elif self._state == 'sb':
                self._subneg_option = byte
                self._subneg_data.clear()
                self._state = 'sb_data'
            elif self._state == 'sb_data':
                if byte == TelnetCmd.IAC:
                    self._state = 'sb_iac'
                else:
                    self._subneg_data.append(byte)
            elif self._state == 'sb_iac':
                if byte == TelnetCmd.SE:
                    commands.append(TelnetCommand(
                        TelnetCmd.SB,
                        self._subneg_option,
                        bytes(self._subneg_data)
                    ))
                    self._state = 'normal'
                elif byte == TelnetCmd.IAC:
                    self._subneg_data.append(TelnetCmd.IAC)
                    self._state = 'sb_data'
                else:
                    self._state = 'normal'

        return bytes(output), commands


class TelnetNegotiator:
    """Handle telnet option negotiation from server side."""

    def __init__(self):
        self.local_options = set()  # Options we WILL do
        self.remote_options = set()  # Options we want client to DO

    def initial_negotiation(self) -> bytes:
        """Send initial negotiation to client."""
        response = bytearray()

        # Server will echo
        response.extend([TelnetCmd.IAC, TelnetCmd.WILL, TelnetOpt.ECHO])
        self.local_options.add(TelnetOpt.ECHO)

        # Server will suppress go-ahead
        response.extend([TelnetCmd.IAC, TelnetCmd.WILL, TelnetOpt.SUPPRESS_GO_AHEAD])
        self.local_options.add(TelnetOpt.SUPPRESS_GO_AHEAD)

        # Request client to send NAWS
        response.extend([TelnetCmd.IAC, TelnetCmd.DO, TelnetOpt.NAWS])

        return bytes(response)

    def handle_command(self, cmd: TelnetCommand) -> bytes:
        """Handle a telnet command, return response bytes."""
        response = bytearray()

        if cmd.cmd == TelnetCmd.WILL:
            # Client offers to do something
            if cmd.option in (TelnetOpt.NAWS, TelnetOpt.TERMINAL_TYPE):
                response.extend([TelnetCmd.IAC, TelnetCmd.DO, cmd.option])
                self.remote_options.add(cmd.option)
            else:
                response.extend([TelnetCmd.IAC, TelnetCmd.DONT, cmd.option])

        elif cmd.cmd == TelnetCmd.WONT:
            # Client refuses
            response.extend([TelnetCmd.IAC, TelnetCmd.DONT, cmd.option])
            self.remote_options.discard(cmd.option)

        elif cmd.cmd == TelnetCmd.DO:
            # Client requests us to do something
            if cmd.option in (TelnetOpt.ECHO, TelnetOpt.SUPPRESS_GO_AHEAD):
                if cmd.option not in self.local_options:
                    response.extend([TelnetCmd.IAC, TelnetCmd.WILL, cmd.option])
                    self.local_options.add(cmd.option)
            else:
                response.extend([TelnetCmd.IAC, TelnetCmd.WONT, cmd.option])

        elif cmd.cmd == TelnetCmd.DONT:
            # Client demands we stop something
            if cmd.option in self.local_options:
                response.extend([TelnetCmd.IAC, TelnetCmd.WONT, cmd.option])
                self.local_options.discard(cmd.option)

        elif cmd.cmd == TelnetCmd.SB:
            # Subnegotiation
            if cmd.option == TelnetOpt.NAWS and cmd.data and len(cmd.data) >= 4:
                # Window size: width(2) + height(2)
                width = (cmd.data[0] << 8) | cmd.data[1]
                height = (cmd.data[2] << 8) | cmd.data[3]
                # Store for pagination if needed
                self._window_size = (width, height)

        return bytes(response)

    @property
    def window_size(self) -> Tuple[int, int]:
        """Get negotiated window size (width, height)."""
        return getattr(self, '_window_size', (80, 24))

File: src-tauri/src/telnet.rs
use crate::session::{SessionConfig, SessionError, SessionHandle, SessionManager, SessionState};
use std::sync::Arc;
use tauri::Emitter;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::sync::mpsc;
use tracing::{debug, error, info, warn};

// Telnet protocol constants
const IAC: u8 = 255; // Interpret As Command
const DONT: u8 = 254;
const DO: u8 = 253;
const WONT: u8 = 252;
const WILL: u8 = 251;
const SB: u8 = 250; // Sub-negotiation Begin
const SE: u8 = 240; // Sub-negotiation End

// Telnet options
const OPT_ECHO: u8 = 1;
const OPT_SUPPRESS_GO_AHEAD: u8 = 3;
const OPT_TERMINAL_TYPE: u8 = 24;
const OPT_NAWS: u8 = 31; // Negotiate About Window Size

struct TelnetParser {
    state: TelnetParseState,
    subneg_option: u8,
    subneg_data: Vec<u8>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
enum TelnetParseState {
    Normal,
    Iac,
    Will,
    Wont,
    Do,
    Dont,
    Sb,
    SbData,
    SbIac,
}

impl TelnetParser {
    fn new() -> Self {
        Self {
            state: TelnetParseState::Normal,
            subneg_option: 0,
            subneg_data: Vec::new(),
        }
    }

    fn parse(&mut self, input: &[u8]) -> (Vec<u8>, Vec<TelnetCommand>) {
        let mut output = Vec::with_capacity(input.len());
        let mut commands = Vec::new();

        for &byte in input {
            match self.state {
                TelnetParseState::Normal => {
                    if byte == IAC {
                        self.state = TelnetParseState::Iac;
                    } else {
                        output.push(byte);
                    }
                }
                TelnetParseState::Iac => match byte {
                    IAC => {
                        output.push(IAC);
                        self.state = TelnetParseState::Normal;
                    }
                    WILL => self.state = TelnetParseState::Will,
                    WONT => self.state = TelnetParseState::Wont,
                    DO => self.state = TelnetParseState::Do,
                    DONT => self.state = TelnetParseState::Dont,
                    SB => self.state = TelnetParseState::Sb,
                    SE => self.state = TelnetParseState::Normal,
                    _ => self.state = TelnetParseState::Normal,
                },
                TelnetParseState::Will => {
                    commands.push(TelnetCommand::Will(byte));
                    self.state = TelnetParseState::Normal;
                }
                TelnetParseState::Wont => {
                    commands.push(TelnetCommand::Wont(byte));
                    self.state = TelnetParseState::Normal;
                }
                TelnetParseState::Do => {
                    commands.push(TelnetCommand::Do(byte));
                    self.state = TelnetParseState::Normal;
                }
                TelnetParseState::Dont => {
                    commands.push(TelnetCommand::Dont(byte));
                    self.state = TelnetParseState::Normal;
                }
                TelnetParseState::Sb => {
                    self.subneg_option = byte;
                    self.subneg_data.clear();
                    self.state = TelnetParseState::SbData;
                }
                TelnetParseState::SbData => {
                    if byte == IAC {
                        self.state = TelnetParseState::SbIac;
                    } else {
                        self.subneg_data.push(byte);
                    }
                }
                TelnetParseState::SbIac => {
                    if byte == SE {
                        commands.push(TelnetCommand::Subnegotiation(
                            self.subneg_option,
                            std::mem::take(&mut self.subneg_data),
                        ));
                        self.state = TelnetParseState::Normal;
                    } else if byte == IAC {
                        self.subneg_data.push(IAC);
                        self.state = TelnetParseState::SbData;
                    } else {
                        self.state = TelnetParseState::Normal;
                    }
                }
            }
        }

        (output, commands)
    }
}

#[derive(Debug)]
#[allow(dead_code)]
enum TelnetCommand {
    Will(u8),
    Wont(u8),
    Do(u8),
    Dont(u8),
    Subnegotiation(u8, Vec<u8>),
}

fn build_response(commands: &[TelnetCommand], cols: u32, rows: u32) -> Vec<u8> {
    let mut response = Vec::new();

    for cmd in commands {
        match cmd {
            TelnetCommand::Will(opt) => {
                // Acknowledge WILL for options we support
                match *opt {
                    OPT_ECHO | OPT_SUPPRESS_GO_AHEAD => {
                        response.extend_from_slice(&[IAC, DO, *opt]);
                    }
                    _ => {
                        response.extend_from_slice(&[IAC, DONT, *opt]);
                    }
                }
            }
            TelnetCommand::Do(opt) => {
                // Handle DO requests
                match *opt {
                    OPT_TERMINAL_TYPE => {
                        response.extend_from_slice(&[IAC, WILL, OPT_TERMINAL_TYPE]);
                    }
                    OPT_NAWS => {
                        // Agree to NAWS and send window size
                        response.extend_from_slice(&[IAC, WILL, OPT_NAWS]);
                        response.extend_from_slice(&build_naws(cols, rows));
                    }
                    OPT_SUPPRESS_GO_AHEAD => {
                        response.extend_from_slice(&[IAC, WILL, OPT_SUPPRESS_GO_AHEAD]);
                    }
                    _ => {
                        response.extend_from_slice(&[IAC, WONT, *opt]);
                    }
                }
            }
            TelnetCommand::Subnegotiation(opt, data) => {
                if *opt == OPT_TERMINAL_TYPE && !data.is_empty() && data[0] == 1 {
                    // Terminal type request (SEND)
                    response.extend_from_slice(&[
                        IAC,
                        SB,
                        OPT_TERMINAL_TYPE,
                        0, // IS
                    ]);
                    response.extend_from_slice(b"xterm-256color");
                    response.extend_from_slice(&[IAC, SE]);
                }
            }
            _ => {}
        }
    }

    response
}

fn build_naws(cols: u32, rows: u32) -> Vec<u8> {
    let cols = cols as u16;
    let rows = rows as u16;
    let mut naws = vec![IAC, SB, OPT_NAWS];

    // Width (2 bytes, big endian)
    let width_hi = (cols >> 8) as u8;
    let width_lo = (cols & 0xFF) as u8;
    if width_hi == IAC {
        naws.push(IAC);
    }
    naws.push(width_hi);
    if width_lo == IAC {
        naws.push(IAC);
    }
    naws.push(width_lo);

    // Height (2 bytes, big endian)
    let height_hi = (rows >> 8) as u8;
    let height_lo = (rows & 0xFF) as u8;
    if height_hi == IAC {
        naws.push(IAC);
    }
    naws.push(height_hi);
    if height_lo == IAC {
        naws.push(IAC);
    }
    naws.push(height_lo);

    naws.extend_from_slice(&[IAC, SE]);
    naws
}

pub async fn run_telnet_session(
    session_id: String,
    config: SessionConfig,
    manager: Arc<SessionManager>,
) -> Result<(), SessionError> {
    let app_handle = manager.app_handle().clone();

    // Create channels for communication
    let (input_tx, mut input_rx) = mpsc::channel::<Vec<u8>>(256);
    let (shutdown_tx, mut shutdown_rx) = mpsc::channel::<()>(1);
    let (resize_tx, mut resize_rx) = mpsc::channel::<(u32, u32)>(16);

    // Store session handle
    let handle = SessionHandle {
        id: session_id.clone(),
        config: config.clone(),
        state: SessionState::Connecting,
        input_tx,
        shutdown_tx,
        resize_tx,
    };
    manager.insert(handle);

    // Emit connecting state
    emit_state(&app_handle, &session_id, SessionState::Connecting);

    // Connect to server
    let addr = format!("{}:{}", config.host, config.port);
    info!(session_id = %session_id, addr = %addr, "Connecting to Telnet server");

    let stream = match TcpStream::connect(&addr).await {
        Ok(stream) => stream,
        Err(e) => {
            error!(session_id = %session_id, error = %e, "Telnet connection failed");
            emit_state(&app_handle, &session_id, SessionState::Error);
            manager.remove(&session_id);
            return Err(SessionError::ConnectionFailed(e.to_string()));
        }
    };

    emit_state(&app_handle, &session_id, SessionState::Connected);
    emit_state(&app_handle, &session_id, SessionState::Ready);
    info!(session_id = %session_id, "Telnet session ready");

    let (mut reader, mut writer) = stream.into_split();
    let mut parser = TelnetParser::new();
    let mut read_buf = [0u8; 4096];
    let mut current_cols = config.cols;
    let mut current_rows = config.rows;

    loop {
        tokio::select! {
            // Read from server
            result = reader.read(&mut read_buf) => {
                match result {
                    Ok(0) => {
                        info!(session_id = %session_id, "Server closed connection");
                        break;
                    }
                    Ok(n) => {
                        let (data, commands) = parser.parse(&read_buf[..n]);

                        // Handle telnet commands
                        if !commands.is_empty() {
                            let response = build_response(&commands, current_cols, current_rows);
                            if !response.is_empty() {
                                if let Err(e) = writer.write_all(&response).await {
                                    warn!(session_id = %session_id, error = %e, "Failed to send telnet response");
                                }
                            }
                        }

                        // Forward clean data to frontend
                        if !data.is_empty() {
                            let event_name = format!("session:{}", session_id);
                            debug!(session_id = %session_id, bytes = data.len(), "Received data from Telnet");
                            if let Err(e) = app_handle.emit(&event_name, data) {
                                error!(session_id = %session_id, error = %e, "Failed to emit data event");
                            }
                        }
                    }
                    Err(e) => {
                        error!(session_id = %session_id, error = %e, "Read error");
                        break;
                    }
                }
            }

            // Handle input from frontend
            Some(data) = input_rx.recv() => {
                debug!(session_id = %session_id, bytes = data.len(), "Sending data to Telnet");
                if let Err(e) = writer.write_all(&data).await {
                    error!(session_id = %session_id, error = %e, "Failed to send data");
                    break;
                }
            }

            // Handle resize requests
            Some((cols, rows)) = resize_rx.recv() => {
                debug!(session_id = %session_id, cols = cols, rows = rows, "Resizing terminal");
                current_cols = cols;
                current_rows = rows;
                let naws = build_naws(cols, rows);
                if let Err(e) = writer.write_all(&naws).await {
                    warn!(session_id = %session_id, error = %e, "Failed to send NAWS");
                }
            }

            // Handle shutdown request
            _ = shutdown_rx.recv() => {
                info!(session_id = %session_id, "Shutdown requested");
                break;
            }
        }
    }

    // Cleanup
    info!(session_id = %session_id, "Telnet session ending");
    emit_state(&app_handle, &session_id, SessionState::Disconnected);
    manager.remove(&session_id);

    Ok(())
}

fn emit_state(app_handle: &tauri::AppHandle, session_id: &str, state: SessionState) {
    let event_name = format!("session:{}:state", session_id);
    if let Err(e) = app_handle.emit(&event_name, state) {
        error!(
            session_id = %session_id,
            error = %e,
            "Failed to emit state event"
        );
    }
}
